tools/oesign/main.c

```diff
@@ -743,23 +830,42 @@ int sign_parser(int argc, const char* argv[])
         }
     } while (1);
 
-    if (enclave == NULL)
+    if (conffile == NULL)
     {
-        Err("Enclave image flag is missing");
+        Err("Config file flag is missing");
         ret = 1;
     }
-    if (conffile == NULL)
+
+#if HAS_ENGINE_SUPPORT
+    if (keyfile && engine_id)
     {
-        Err("Config file flag is missing");
+        Err("only one of keyfile and engine flags are allowed");
         ret = 1;
+        goto done;
+    }
+    if ((keyfile == NULL && engine_id == NULL) && engine_load_path != NULL)
```

Comment:
The error message doesn't match the logic of this conditional; I'm not sure why you're checking that the `engine_load_path` is non-null in this case. This is also not logically equivalent to the checks I was asking for in the previous iteration:

> Additional error cases to trap:
> - if (engine_id != NULL && key_id == NULL) 
> - if (engine_id == NULL && (key_id != NULL || engine_load_path != NULL))

If you want to combine all the checks, they boil down to:

```c
if ((keyfile == NULL) == (engine_id == NULL))
{
    // You must specify either the keyfile or engine_id parameter, but not both.
}
if ((engine_id != NULL) && (key_id == NULL))
{
    // You must specify a key_id if providing an engine_id
}
if ((engine_id == NULL) && (key_id != NULL || engine_load_path != NULL))
{
    // You must not specify a key_id or engine_load_path if not using engine_id (i.e. using keyfile)
}
```